### PR TITLE
Environment variable iterator overrides

### PR DESCRIPTION
**Pull request by Nemo157**
For case #24214, additional iterators mentioned in https://github.com/rust-lang/rust/issues/24214#issuecomment-98555383.

Overrides `std::env::{ Vars, VarsOs }::{ count, nth, last}` to proxy through to their inner iterator (would be nice if there were some iterator wrapper trait that meant these didn't need to be implemented every time an iterator was wrapped...).

Overrides `std::sys::unix::os::Env::{ count, nth, last }` to proxy through to the inner `vec::IntoIter`. Should make `count` faster, and future proofs in case `last` or `nth` get optimised implementations in the future.

Overrides `std::sys::windows::Env::{ count, nth, last }` to not allocate `OsString`s when not needed. Done by defining a sub-iterator that returns the raw pointers and lengths, this uses the default implementations as they should be pretty much equivalent to manually implemented versions, then in the outer iterator does the allocation with the result.

I don't have a machine I can actually test the windows changes on, with some hackery (https://gist.github.com/Nemo157/0d1f02b08c132f1a00da) I managed to get the type checker at least running to ensure it should compile. It would be nice if there were some way to test the higher level platform specific code. As far as I can tell testing `windows::Env` on unix would work fine if the infrastructure was setup for it.

### CODE REVIEW

File: src/libstd/sys/windows/os.rs [hunk 1]
```diff
@@ -72,14 +72,18 @@ pub fn error_string(errnum: i32) -> String {
 }
 
 pub struct Env {
+    inner: EnvSlices,
+}
+
+struct EnvSlices {
```
Comment (alexcrichton) [inline src/libstd/sys/windows/os.rs]: To me when I initially read this it wasn't clear why the separation here was needed, could you add a comment explaining why these are separate?

File: src/libstd/sys/windows/os.rs [hunk 2]
```diff
@@ -100,16 +104,41 @@ impl Iterator for Env {
                     Some(p) => p,
                     None => continue,
                 };
-                return Some((
-                    OsStringExt::from_wide(&s[..pos]),
-                    OsStringExt::from_wide(&s[pos+1..]),
-                ))
+                return Some((p, pos, len as usize))
```
Comment (alexcrichton) [inline src/libstd/sys/windows/os.rs]: Could this return `s`? That seems like it may be tad bit safer than returning raw pointers

**Comment by alexcrichton** [comment]
Could you add some tests as well which exercise these functions?

**Comment by Nemo157** [comment]
@alexcrichton done.

Again I could only type check the tests as I don't have a windows machine setup for rust dev.

I hope using the type alias for `sys::windows::os::Env` is ok? I couldn't figure out any other way to make sure the tests had the correct lifetime and use slices between the iterators without either having to add a unused lifetime parameter to the `sys::unix::os::Env` or change the public API.

**Comment by alexcrichton** [comment]
Ah that may be an extra layer or two of indirection which may not be worth it, perhaps just `Env` could be used which uses the inner copy with `'static`? I'd also be fine if `EnvSlices` just returned `&'static [u16]` and documenting that it's unsafe and the lifetime needs to be contained elsewhere.

Also, could you move the tests so they cover both the Unix and Windows cases? That should also help exercise the implementation in terms of what's actually being run with `env::vars`. I'd recommend adding the test to `run-pass` to ensure it's isolated from everything else going on.

File: src/test/run-pass/env-vars-iter.rs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                assert_eq!(vars.next(), None);
+            },
+            arg => {
+                panic!("Unexpected arg {}", arg)
+            },
+        }
+    } else {
+      let status = Command::new(env::current_exe().unwrap())
+                           .arg("empty")
+                           .env_clear()
```
Comment (alexcrichton) [inline src/test/run-pass/env-vars-iter.rs]: This may or may not pass on Windows due to https://github.com/rust-lang/rust/issues/31259

Comment (Nemo157) [inline src/test/run-pass/env-vars-iter.rs]: Damn, that is... annoying :frowning:

File: src/test/run-pass/env-vars-iter.rs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    } else {
+        // Command::env_clear does not work on Windows.
+        // https://github.com/rust-lang/rust/issues/31259
+        if !cfg!(windows) {
+          let status = Command::new(env::current_exe().unwrap())
+                               .arg("empty")
+                               .env_clear()
+                               .status()
+                               .unwrap();
+          assert_eq!(status.code(), Some(0));
```
Comment (alexcrichton) [inline src/test/run-pass/env-vars-iter.rs]: Ah looks like the tabs are a little off here (2 instead of 4 spaces)

**Comment by alexcrichton** [comment]
Ok, this looks good to me, could you also squash the commits together?

**Comment by alexcrichton** [comment]
@bors: r+ 52d506e4821f949f4a64cf7ba892ac87229d7ae6

Thanks!

**Comment by alexcrichton** [comment]
Closing due to inactivity, but feel free to reopen with tests fixed!

**Pull request closed without merging**